external-crates/move/crates/move-compiler/src/cfgir/optimize/mod.rs

```diff
@@ -64,12 +72,34 @@ pub fn optimize(
             debug_assert_eq!(count, opt_count);
             break;
         }
-
         // reset the count if something has changed
         if optimization(signature, locals, constants, cfg) {
             count = 0
         } else {
             count += 1
         }
     }
+    // If we aren't optimizing a constant, attempt to coalesce locals and re-optimize.
+    if !is_constant && optimize_2024 {
+        let new_locals = coalesce_locals::optimize(
+            signature,
+            infinite_loop_starts,
+            locals.clone(),
+            cfg,
+        );
+        *locals = new_locals;
+        cfg.recompute();
+
+        // Re-optimize, but don't recur again.
+        optimize(
+            env,
+            package,
+            signature,
+            infinite_loop_starts,
+            locals,
+            constants,
+            cfg,
+            true,
+        );
```

Comment:
Maybe we should break the loop out into its own function so we don't have to control this access via flag